### PR TITLE
Gt/hex grid args

### PR DESCRIPTION
**Pull request by gregtucker**
This PR adds a keyword argument "node_layout" to HexModelGrid. This new keyword replaces "shape", which can now be repurposed as a 2-element tuple with (rows, cols), consistent with RasterModelGrid and Numpy usage. This change implements the following behavior that maintains backward compatibility as follows:

1) If user sends shape as a str ('rect' or 'hex'), it is interpreted as node_layout, and a deprecation warning is given.
2) If user sends shape as a type, it is interpreted as (base_num_rows, base_num_cols) (and overrides the corresponding positional arguments.

Note that this PR does NOT handle the case of a user sending a tuple as the first positional argument. In other words, if they want to pass tuple, they have to use the (now modified) shape keyword.

This PR addresses most of issue #864

### CODE REVIEW

**Comment by gregtucker** [comment]
(note typo above: "type" = "tuple")

**Comment by kbarnhart** [comment]
@gregtucker awesome will take a closer look!

@gregtucker & @mcflugen since this changes the meaning of `shape` I think that this means this is the first PR that needs to go into our soon-to-be-created LL2.0 branch?

If so, then we need another branch that JUST makes the name change and yields a deprecation warning... (but doesn't repurpose shape).

**Comment by gregtucker** [comment]
Thanks @kbarnhart. Actually this PR is for the current LL, and should be fully backward compatible. 

LL 2.0 will need to add a shift in the positional args, from (NR, NC, dx, ...) to ((NR, NC), dx, ...)

**Comment by gregtucker** [comment]
To be clear, you can still do 

shape='rect' (or 'hex')

and it will be interpreted correctly (but send a deprecation warning). You can alternative do

shape=(nrows, ncols)

and that too will be interpreted correctly, as will

node_layout='rect' (or 'hex')

File: landlab/grid/hex.py [hunk 1]
```diff
@@ -118,6 +119,14 @@ def __init__(
...
         7
         """
+
+        # Pre-LL2.0: handle shape and node_layout keyword args
+        # (note: in LL2.0, shape should refer ONLY to rows and columns)
+        if type(shape) is str:  # "old" (LL < 2.0) usage
+            node_layout = shape
+        elif type(shape) is tuple:
+            (base_num_rows, base_num_cols) = shape
+
```
Comment (kbarnhart) [inline landlab/grid/hex.py]: @gregtucker I think if we want this to be in current LL (as opposed to LL2.0) we need to not have lines 127 and 128... Doesn't that permit someone to pass shape as tuple and then have it used as the nr and nc?

**Comment by kbarnhart** [comment]
@gregtucker yes, the functionality you pointed out is what I think needs to change. 

It is backward compatible...but also forward compatible. Since it is the same keyword argument this seems not-correct. 

But perhaps you or @mcflugen  have better founded opinions on how to implement this sort of backward-forward compatibility...

File: landlab/grid/hex.py (hunk 1 above)
Comment (gregtucker) [reply landlab/grid/hex.py]: Yes, but I don't think that's a problem. It supports both old usage, when shape is a str, and new usage, when shape is a tuple. In other words, it should still be fully compatible with old usage, i.e., is still backward compatible.

File: landlab/grid/hex.py [hunk 2]
```diff
@@ -235,24 +246,33 @@ def _initialize(
...
 
+        # Pre-LL2.0: handle shape and node_layout keyword args
+        # (note: in LL2.0, shape should refer ONLY to rows and columns)
+        if type(shape) is str:  # "old" (LL < 2.0) usage
+            node_layout = shape
+            warn(message="Use node_layout to specify 'rect' or 'hex' layout",
+                 category=DeprecationWarning)
+        elif type(shape) is tuple:
+            (base_num_rows, base_num_cols) = shape
+
```
Comment (kbarnhart) [inline landlab/grid/hex.py]: @gregtucker my only comments are to:

- [ ] Add a statement to the deprecation warning that indicates that this change will be enforced in LL2.0, and that shape will have a new meaning in LL2.0
- [ ] Add a test that the deprecation warning is raised correctly.
- [ ] Add a test that if shape is passed a tuple, it behaves as expected. 

I couldn't find tests in the PR, and I double checked that [coveralls indicates](https://coveralls.io/builds/22923077/source?filename=landlab/grid/hex.py) that those lines are still untested.

Comment (gregtucker) [reply landlab/grid/hex.py]: Good ideas, thanks---now added.

**Pull request merged**